signac/core/mongodbcollection.py

```diff
@@ -0,0 +1,147 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Implements Mongo-backend.
+
+This implements the Mongo-backend for SyncedCollection API by
+implementing sync and load methods.
+"""
+from .synced_collection import SyncedCollection
+from .syncedattrdict import SyncedAttrDict
+from .synced_list import SyncedList
+
+
+class MongoCollection(SyncedCollection):
+    """Implement sync and load using a Mongo backend."""
+
+    backend = __name__  # type: ignore
+
+    def __init__(self, name=None, client=None, database='signac_db', collection='collection',
+                 mongo_kwargs=None, **kwargs):
+        import pymongo
+
+        if client is None:
+            mongo_kwargs = mongo_kwargs if mongo_kwargs is not None else {}
+            self._client = pymongo.MongoClient(**mongo_kwargs)
+        else:
+            self._client = client
+        self._db = self._client.get_database(database)
+        self._collection = self._db.get_collection(collection)
+        self._name = name
+        self._key = type(self).__name__ + '::name'
+        super().__init__(**kwargs)
+        if (name is None) == (self._parent is None):
+            raise ValueError(
+                "Illegal argument combination, one of the two arguments, "
+                "parent or name must be None, but not both.")
```

Comment:
We can massively simplify this constructor implementation and improve the API by only accepting an already instantiated instance of [`pymongo.Collection`](https://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection) here. This has the benefit that we don't need to sanitize and validate the input arguments, deal with establishing the connection and related issues, and we don't need to import `pymongo` at all within this module. Finally, this API would be much more flexible for users and developers and decouples it from potential breaking upstream changes in the `pymongo` package.
```suggestion
    def __init__(self, collection, **kwargs):
        self._collection = collection
        super().__init__(**kwargs)
```